cli/src/main/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/package.scala

```diff
@@ -16,13 +16,108 @@
+   * `kubernetesArgs` are supplied, the function is invoked and the resulting [[Try[Done]] is returned.
+   */
+  def generateResources(getDockerConfig: String => Try[Config], outputHandler: Seq[GeneratedKubernetesResource] => Unit)(generateDeploymentArgs: GenerateDeploymentArgs, kubernetesArgs: KubernetesArgs): Try[Done] = {
+
+    def getLabel(config: Config): Try[Map[String, String]] =
+      config.config.Labels match {
+        case Some(v) if v.nonEmpty => Success(v)
+        case _ => Failure(new IllegalArgumentException("Unable to generate Kubernetes resources from empty labels."))
+      }
+
+    def generateIngress(annotations: Annotations): Try[Seq[GeneratedKubernetesResource]] =
+      kubernetesArgs.ingressArgs match {
+        case Some(IngressIstioArgs) => IngressIstio.generate(annotations).map(Seq(_))
+        case Some(v: IngressNgnixArgs) => IngressNginx.generate(annotations, v.tlsSecretName, v.sslRedirect).map(Seq(_))
+        case _ => Success(Seq.empty)
+      }
+
+    for {
+      config <- getDockerConfig(generateDeploymentArgs.dockerImage.get)
+
+      label <- getLabel(config)
+
+      annotations = Annotations(label, generateDeploymentArgs)
+
+      deploymentJson <- Deployment.generate(
+        annotations,
+        kubernetesArgs.kubernetesVersion.get,
+        generateDeploymentArgs.dockerImage.get,
+        kubernetesArgs.deploymentArgs.imagePullPolicy,
+        kubernetesArgs.deploymentArgs.numberOfReplicas)
+
+      serviceJson <- Service.generate(annotations, kubernetesArgs.serviceArgs.clusterIp)
+
+      ingressJson <- generateIngress(annotations)
+    } yield {
+      outputHandler(Seq(deploymentJson, serviceJson) ++ ingressJson)
+      Done
+    }
+  }
+
+  /**
+   * Accepts the instructions supplied by the `output`, and returns the appropriate function to handle the
+   * generated Kubernetes resources.
+   */
+  def handleGeneratedResources(output: KubernetesArgs.Output): Seq[GeneratedKubernetesResource] => Unit =
+    output match {
+      case KubernetesArgs.Output.PipeToKubeCtl(out) => pipeToKubectl(out)
+      case KubernetesArgs.Output.SaveToFile(path) => saveToFile(path)
+    }
+
+  private[kubernetes] def saveToFile(path: Path)(generatedResources: Seq[GeneratedKubernetesResource]): Unit = {
+    logger.info(s"Saving to ${path.toAbsolutePath}")
+    if (!Files.exists(path)) {
+      Files.createDirectories(path)
+    }
+
+    generatedResources.foreach { r =>
+      val fileName = s"${r.resourceType}-${r.name}.json"
+      val file = path.resolve(fileName)
+      val formattedJson = r.payload.pretty(PrettyParams.spaces2.copy(preserveOrder = true))
+
+      logger.info(fileName)
+      Files.deleteIfExists(file)
+      Files.write(file, formattedJson.getBytes)
+    }
+
+    logger.info("Done!")
+  }
+
+  private[kubernetes] def pipeToKubectl(out: PrintStream)(generatedResources: Seq[GeneratedKubernetesResource]): Unit =
```

Comment:
Nitpick - it's just piping to stdout, so maybe rename to indicate that?

Reply:
I've renamed it to `pipeToStream` - wdyt?

The `out` is not necessarily `stdout` since in one of the unit test I've wrapped `ByteArrayOutputStream` inside `PrintStream` so I can assert the output.

Reply:
:+1: